OpenSim/Analyses/OutputReporter.cpp

```diff
@@ -0,0 +1,136 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.   *
+ * See the License for the specific language governing permissions and        *
+ * limitations under the License.                                             *
+ * -------------------------------------------------------------------------- */
+
+
+//=============================================================================
+// INCLUDES
+//=============================================================================
+#include "OutputReporter.h"
+#include <OpenSim/Simulation/Model/Model.h>
+#include <OpenSim/Common/IO.h>
+
+using namespace OpenSim;
+using namespace std;
+
+//=============================================================================
+// ANALYSIS
+//=============================================================================
+int OutputReporter::begin(SimTK::State& s)
+{
+    if (!proceed()) return(0);
+
+    _pvtModel.reset(_model->clone());
+
+    _tableReporterDouble = new TableReporter_<double>();
+    _tableReporterVec3 = new TableReporter_<SimTK::Vec3>();
+    _tableReporterSpatialVec = new TableReporter_<SimTK::SpatialVec>();
+
+    _pvtModel->addComponent(_tableReporterDouble.get());
+    _pvtModel->addComponent(_tableReporterVec3.get());
+    _pvtModel->addComponent(_tableReporterSpatialVec.get());
+
+    for (int i = 0; i < getProperty_output_paths().size(); ++i) {
+        auto outpath = ComponentPath(get_output_paths(i));
+        // assume that Outputs are for the Model
+        string compName = _pvtModel->getAbsolutePathString();
+        // if a a ComponentPath is specified then use the parent path
+        if (outpath.getNumPathLevels()) {
+            compName = outpath.getParentPathString();
+        }
+        auto outputName = outpath.getComponentName();
+
+        // In keeping with assumption of Outputs for the Model (Component)
+        const Component* comp = _pvtModel.get();
+
+        // If the ComponentPath has a Component name get that Component
+        if (!compName.empty())
+            comp = &_pvtModel->getComponent(compName);
+
+        auto& out = comp->getOutput(outputName);
+        if (out.getTypeName() == "double") {
+            _tableReporterDouble->addToReport(out);
+        }
+        else if (out.getTypeName() == "Vec3") {
+            _tableReporterVec3->addToReport(out);
+        }
+        else if (out.getTypeName() == "SpatialVec") {
+            _tableReporterSpatialVec->addToReport(out);
+        }
+        else {
+            cout << "Output '" << out.getPathName() << "' of type "
+                << out.getTypeName() << " is not supported by OutputReporter."
+                << " Consider adding a TableReporter_ to the Model." << endl;
+        }
+    }
+
+    _pvtModel->initSystem();
+
+    return 0;
+}
+
+int OutputReporter::step(const SimTK::State& s, int stepNumber)
+{
+    if (!proceed(stepNumber)) {
+        return 0;
+    }
+    if (_pvtModel == nullptr)
+        return -1;
+
```

Comment:
Is it possible for _pvtModel to be nullptr at all?

Reply:
Would there be an issue with realizing the state to Report using the cloned model having unexpected side effects (e.g. if model has no dynamics/masses, will that throw or should Outputs realize lazily on their own to the needed stage?)

Reply:
> Is it possible for _pvtModel to be nullptr at all?

There is nothing stopping anyone from calling Analysis::step() AFAIK. Other analyses had similar check for `_model == NULL` so I figured to leave it in. It does make more sense for the check to be on `_model` at begin(), so I'll move it there.

Reply:
> Would there be an issue with realizing the state to Report using the cloned model having unexpected side effects (e.g. if model has no dynamics/masses, will that throw or should Outputs realize lazily on their own to the needed stage?)

Good point. Yes, definitely there is an issue with realizing to `Report` when `Dynamics` and `Acceleration` are invalid. I am not going to access `Output`s here myself because I want to leverage the `TableReporter` to do the actual reporting. I might be able to get the `TableReporter` to report *on demand*, however. I'll look into it. Thanks.